README.rst
.. contents:: **Table of contents**

Introduction
============

Scope of this product is to make possible an expand/collapse feature in Plone navigation(s)
portlet without any modification to the navigation code itself, or any needs to override its features.

This product *is not* a new navigation portlet, is just a JavaScript add-on that rely on native Plone's
`jQuery`__ support.

__ http://jquery.com/

More important: it is only focused to make *some* navigation links to be expandable. 

When you will like this?
------------------------

The collective.navigationtoggle is useful when your Plone site needs (not much) special handling of
navigation elements. Sometimes your site structure is someway like this::

    ROOT
    |
    |_ FolderWhatever
    \_ NotImportantFolder
       |
       |_ ImportantFolder1
       ...
       \_ ImportantFolderN

If you configure your navigation portlet on the site root, users must click on the *NotImportantFolder*
and reload the whole page to reach the *ImportantFolder*'s section.
The site at the *NotImportantFolder* level is not useful... maybe you only give a *folder_listing* view
or a default page that only say "Welcome to an important area of the site, please visit the subsection
you are looking for...".

For the user experience the first click is only a waste of time.

What you are looking for can be a client side effect that expand/collapse the folder in the navigation.
The *NotImportantFolder* itself is not seen as a real content in your information architecture.

What will change
----------------

.. image:: http://blog.redturtle.it/pypi-images/collective.navigationtoggle/collective.navigationtoggle-0.3.0-01.png/image_mini
   :align: left
   :alt: Closed navigation from AUSL site
   :target: http://blog.redturtle.it/pypi-images/collective.navigationtoggle/collective.navigationtoggle-0.3.0-01.png

.. image:: http://blog.redturtle.it/pypi-images/collective.navigationtoggle/collective.navigationtoggle-0.3.0-02.png/image_mini
   :align: right
   :alt: Open navigation from AUSL site
   :target: http://blog.redturtle.it/pypi-images/collective.navigationtoggle/collective.navigationtoggle-0.3.0-02.png

Make possible that special navigation links will no more move the user to the target section but simply
shows in the navigation itself all subsections (so the navigation seems like the user really moved to
the target folder).

A second click will collapse the section.

Default page in a folder, elements marked with "*Exclude from navigation*" and unwanted
type from the ``metaTypesNotToList`` property will be excluded.

The script try to simulate best at possible a normal portlet navigation behavior.

The code keeps in mind *graceful degradation*. Browser without JavaScript enabled will simply use basic
Plone navigation features.

Detailed documentation
======================

Basic configuration
-------------------

In your Plone configuration panel you'll find the new "*Navigation Toggle settings*".

.. image:: http://blog.redturtle.it/pypi-images/collective.navigationtoggle/collective.navigationtoggle-0.4.0-01.png/image_preview
   :alt: Setup of Navigation Toggle configuration
   :target: http://blog.redturtle.it/pypi-images/collective.navigationtoggle/collective.navigationtoggle-0.4.0-01.png


From this section you can configure a lot of advanced options, mainly you need to configure:

**Link selectors**
    A set of URL path suffix like ``/foo1/foo2`` of an *href* attribute for a link.
    With this example a link like this will be *hit*::
    
        <a href="http://plonehost/plone/foo1/foo2">
    
    Commonly only link inside navigation portlet are checked (see the "*Toggle container classes*").
    
    You can also provide here a complex jQuery expression, for example::
    
        li.navTreeItem a
    
    If you don't directly provide a selector that match an HTML <A> element, you still need to query for something
    *inside* that kind of node, for example::
    
        li.navTreeItem a.contenttype-folder span
    
    In the last example above, only a click of the navigation entry text will trigger the expand/collapse features,
    while clicking on the icon will keep the default browser behavior.

**Animation delay**
    The product can enable for you a graphical effect when you expand/collapse items. This is disabled by
    default.

    To enable it you must change this value to something higher that 0. The value you put there is the time
    (in millisecond) used for the `slideDown`__ and `slideUp`__ jQuery's effect.

    __ http://api.jquery.com/slideDown/
    __ http://api.jquery.com/slideUp/

    Simply put it to 0 again (the default) to disable effect.

Advanced configurations
-----------------------

**Cache**
    Check it to perform client side cache of the AJAX request.
    If disabled a request to the server will be performed for every click on navigations.

**Toggle container classes**
    A set of CSS classes that mark you navigation elements.
    Only links inside elements with one of those classes are "toggleable".
    
    Note that this option is ignored for "Links selectors" entries where you used a jQuery
    expression and not a URL path.

**HTML list type**
    The type of HTML container element to looks for.

**HTML list item**
    The type of HTML item element to be generated.

Styles
------

This products dynamically adds two new possible CSS classes to ``<li>`` elements that contains links that match
the configuration.

Class "*cnavClosed*" is added when a special navigation elements is shown on the page and when you close
a subtree.
When a subsection is open, the class "*cnavOpen*" is added to the same element.

The generated substructure is a copy of the main structure given by Plone, but a "*cnavGenerated*" class is
added to it.

You can (but this product doesn't) rely on those classes to give additional styles effects in your
Plone theme.

Not standard navigation (...or portlet... or whatever you want)
---------------------------------------------------------------

This products *may* also works on non-standard navigation HTML structure (as the generations on subelement
is done cloning existing node from the same navigation portlet) but some assumptions are done.

As the code works almost client side, you can also use it in portlets that are not standard navigation.
Theoretically you can use it with something that is not even a portlet!

Versions/Dependencies
=====================

Browsers
--------

* Firefox 16 (...probably a new major version will be released while I'm typing this line...)
* Google Chrome 23 (...same as above...)
* Opera 12
* Safari 5.1
* Internet Explorer 8, 9

Plone
-----

* Plone 3.3 (classic Plone theme)
* Plone 4.2 (classic Plone theme and Sunburst)
* Plone 4.3 (classic Plone theme and Sunburst)

Dependencies
------------

* `simplejson`__ (if using Plone 3)
* jQuery 1.3 or better
* `plone.app.registry`__

__ http://pypi.python.org/pypi/simplejson
__ http://pypi.python.org/pypi/plone.app.registry

Cache controls
==============

Toggle configuration are stored in a JavaScript file that your browser and Plone *portal_javascript* tool
will probably cache.

When changing configuration you can use the "*Save and invalidate JS cache*" button.

Other products
==============

If you need a complete expand/collapse navigation feature this product is not for you; so take a
look at:

* `collective.portlet.explore`__
* `Solgema.NavigationPortlet`__

__ http://pypi.python.org/pypi/collective.portlet.explore
__ http://plone.org/products/solgema.navigationportlet

Credits
=======

Developed with the support of:

* `S. Anna Hospital, Ferrara`__
  
  .. image:: http://www.ospfe.it/ospfe-logo.jpg 
     :alt: S. Anna Hospital logo

* `Azienda USL Ferrara`__

  .. image:: http://www.ausl.fe.it/logo_ausl.gif
     :alt: Azienda USL logo

All of them supports the `PloneGov initiative`__.

__ http://www.ospfe.it/
__ http://www.ausl.fe.it/
__ http://www.plonegov.it/

Authors
=======

This product was developed by RedTurtle Technology team.

.. image:: https://www.redturtle.it/redturtle_banner.png
   :alt: RedTurtle Technology Site
   :target: http://www.redturtle.it/

Special thanks to Yang Hong for massive bug fixing.


docs/HISTORY.rst
Changelog
=========

1.0.4 (unreleased)
------------------

- Nothing changed yet.


1.0.3 (2020-05-05)
------------------

- added Transifex.net service integration to manage the translation process [macagua]
- added Spanish translation [macagua]
- updated gettext files support [macagua]

1.0.2 (2014-07-17)
------------------

- fixed a bug in @@query-subelements when using the product in a Plone site that's inside 
  another folder (filesystem, for example) [mpampols]
- fixed a bug that badly put rules inside the generated ``collective.navigationtoggle.rules.js``
  if they contains a quote char [keul]

1.0.1 (2013-01-24)
------------------

- use the object's id as the fallback title for expanded item
  (close `#6`__) [yangh]

  __ https://github.com/RedTurtle/collective.navigationtoggle/issues/6

1.0.0 (2013-01-15)
------------------

- fixed a bug when used with containers with spaces inside the id
  (close `#2`__) [keul]
- file icon of generated elements where not handled in the right way
  (close `#3`__) [keul]
- added a way to invalidate JS cache
  (close `#1`__) [keul]
- now general selectors can match also elements that are not <a>, but
  link's contents
  (close `#4`__) [keul]
- now checking the ``typesUseViewActionInListings`` configuration
  (close `#5`__) [keul]

  __ https://github.com/RedTurtle/collective.navigationtoggle/issues/2
  __ https://github.com/RedTurtle/collective.navigationtoggle/issues/3
  __ https://github.com/RedTurtle/collective.navigationtoggle/issues/1
  __ https://github.com/RedTurtle/collective.navigationtoggle/issues/4
  __ https://github.com/RedTurtle/collective.navigationtoggle/issues/5

0.4.0 (2012-11-15)
------------------

* added a product's browser layer
  [keul]
* provided an own configuration panel
  [keul]
* added all remaining configutation in Plone UI
  [keul]
* ``toggleContainerClass`` now is ``toggleContainerClasses``
  (can be used for a set of classes)
  [keul]
* added i18n support and italian translation
  [keul]
* changes AJAX calling context, now always calling portal root URL
  [keul]
* added tests (both Plone and JavaScript)
  [keul]

0.3.0 (2011-06-07)
------------------

* you can now use more general `jQuery selector`__ instead of simple URL path [keul]
* do not show types that are in list of unwanted types in navigation [keul]
* ``collective.navigationtoggle.txt`` was empty [keul]
* JSLint cleanup [keul]
* code is more general, and new properties likes ``toggleContainerClass``, ``listType``
  and ``listItem`` has been added.
  The code now can be used to simulate navigators also with not portlet navigator
  structures [keul]
* added the uninstall GS step [keul]
* fixed the `simplejson`__ dependency correctly, only if with Python < 2.6 [keul]
* links selectors can be added through Plone UI [keul]

__ http://docs.jquery.com/Selectors
__ http://pypi.python.org/pypi/simplejson

0.2.1 (2010-07-09)
------------------

* fixed a bug when fast double-click on opened elements (close `#1`__) [keul]
* much better Sunburst support (and other themes, in general), now checking also for "*contenttype-xx*"
  and "*state-xx*" CSS classes on links [keul]

__ http://plone.org/products/collective.navigationtoggle/issues/1

0.2.0 (2010-04-11)
------------------

* query_subelement view documentation was incomplete [keul]
* when querying empty folder, do not prevent (in facts: simulate) browser default action [keul]
* prevent browser cache (mainly know Internet Explorer problem) [keul]
* do not always display the content's image icon (sometimes users don't see them in navigation) [keul]
* now every element node cache its subtree, so AJAX call is performed only once (can be disabled) [keul]
* added dependency on simplejson for formatting server responses [keul]
* added minimal Plone 4.0 compatibility (UI result is not perfect) [keul]
* added a simple slide up/down effect (disabled by default) [keul]

0.1.0 (2010-04-06)
------------------

* initial release



collective/navigationtoggle/browser/collective.navigationtoggle.js
/*jslint white: true, undef: true, nomen: true, regexp: true, plusplus: false, bitwise: true, newcap: true, maxerr: 50, vars: true, indent: 4 */
/*global jQuery: false, document: false, window: false */

/**
 * The collective.navigationtoggle Javascript source.
 * Before think "Ehi, I can write this code more and more jQueryish" be aware...
 * Here I'm trying to get the same feature given up by the jQuery.live() method whitout
 * using it (due to some limitation I found).
 */

jQuery.collective_navigationtoggle = {
    /**
     * Store there all references to navigation elements that must be made
     * expansible/collapsible.
     * Every element can be HREF trailer, that will be looked inside every A elements of
     * navigations or a normal jQuery selector.
     */
    toggle_elements: [],
    
    /**
     * Duration in millisecs for control animation slideUp/slideDown effects when
     * expand/collapse navigation elements. Default to 0 (animation disabled)
     */
    slide_animation: 0,
    
    /**
     * Flag for enable or disable client side cache of HttpXmlRequest
     */
    cache: true,
    
    /**
     * Providing those classes, only provided path inside elements with one of those classes are taken.
     * Used only if you provide path(s) in the toggle_elements.
     */
    toggleContainerClasses: ['portletNavigationTree'],
    
    /**
     * What kind of item must contain the link?
     */
    listType: 'ul',
    listItem: 'li',
	
	/**
	 * For testing purpose, just wrap the window.location.href call
	 */
	_goTo: function(url) {
		window.location.href = url;
	}
};

(function ($) {
    $(document).ready(function () {

        /**
         * Context URL to be used for all AJAX call
         */
		/*
        var call_context = $("head base").attr('href');
        if (call_context.charAt(call_context.length - 1) !== '/') {
            call_context = call_context + '/';
        }
        */
		var call_context = portal_url + '/';
    
        /*
         * Don't want to call the context when is in the portal factory. See the Ale's blog post:
         * http://blog.redturtle.it/redturtle-blog/2010/03/11/careful-with-that-ajax-eugene
         */
		if (call_context.indexOf('/portal_factory') > -1) {
            call_context = call_context.substring(0, call_context.indexOf('/portal_factory') + 1);
        }
		
        var loading_time = new Date().getTime();
        
        /**
         * Generate a new navigation element
         * @param {Object} data structure to be used for filling the element
         * @param {jQuery} wrapper a wrapper element to be used as container
         * @param {boolean} withImage include or not the icon image
         * @param {boolean} wrapDiv wrap all generated HTML in a DIV (Plone 3.5/4.0 difference)
         * @param {String} reviewStateClass name on the CSS class related to item review state
         * @param {String} contentTypeClass name on the CSS class related to item content type
         */
        var makeSubelement = function (data, wrapper, withImage, wrapDiv, reviewStateClass, contentTypeClass) {
            var newE = $('<a href="' + data.url + '" title="' + data.description + '">' +
                    (withImage ? '<img alt="' + data.type + '" width="16" height="16" src="' + data.icon + '"/>&nbsp;' : '') +
                    '<span>' + data.title + '</span></a>');
            if (reviewStateClass) {
                newE.addClass("state-" + data.review_state_normalized);
            }
            if (contentTypeClass) {
                newE.addClass("contenttype-" + data.type_normalized);
            }
            return wrapper.append(wrapDiv ? $('<div></div>').append(newE) : newE);
        };
        
        /**
         * Get a class (commonly navTreeLevelX) from a jQuery element and rise X to be X+1.
         * Check are done to be sure that X is an integer. The procedure change the jQuery element
         * inline and don't return anything.
         * @param {jQuery} element the jQuery element on which rise the class
         * @param {String} class the class name (if it isn't navTreeLevelX)
         */
        var riseNavigatorClass = function (element, className) {
            var classes = element.attr('class').split(" "), i;
            if (!className) {
                className = "navTreeLevel";
            }
            for (i = 0; i < classes.length; i++) {
                if (classes[i].indexOf(className) === 0 && classes[i].length === className.length + 1) {
                    var c = classes[i].charAt(classes[i].length - 1);
                    if (!isNaN(c)) {
                        element.removeClass(className + c).addClass(className + (parseInt(c, 10) + 1));
                    }
                }
            }
        };
    
    
        /**
         * Apply to the element the toggle feature
         * @param {Integer} index iteration index
         * @param {Object} value can be part of a link url (like "/foo/foo1") but also a jQuery selector
         */
        var makeDynamicElements = function (index, value) {
            var elements = null;
            
            // try to obtain something from the value as jQuery selector
            try {
                elements = $(value);
                if (elements.length === 0) {
                    elements = null;
                }
            } catch (e) {};
    
            // in that case, only a path was provided
            if (!elements) {
                if ($.collective_navigationtoggle.toggleContainerClasses) {
                    elements = $([]);
                    $.each($.collective_navigationtoggle.toggleContainerClasses, function (index, cssClass) {
                        elements = elements.add("." + cssClass + " a[href$='" + escape(value) + "']");
                    });
                } else {
                    elements = $("a[href$='" + escape(value) + "']");
                }
            }

            elements.each(function () {
                var control = $(this);
                var main_elem = control.closest($.collective_navigationtoggle.listItem);
                if (main_elem.data("cnavMarker")) {
                    return;
                }
                // mark this element to prevent further call to makeDynamicElements
                main_elem.data('cnavMarker', true);
                var wrapDiv = control.parent().is('div'); // For handle Plone3.3 and 4 theme difference
                var ul_model = main_elem.closest($.collective_navigationtoggle.listType).clone(false).addClass('cnavGenerated').empty();
                var a_model = main_elem.find('a');
                var li_model = main_elem.clone(false).empty();
        
                // For themes (like Sunburst) that may add additional classes to elements
                var aModelClasses = []; 
                if (a_model.attr("class")) {
                    aModelClasses = a_model.attr("class").split(" ");
                }
                var aReviewStateClass = null;
                var aContentTypeClass = null;
                $.each(aModelClasses, function (index, value) {
                    var regexpReview = /^state\-/;
                    if (regexpReview.test(value)) {
                        aReviewStateClass = value.replace("state-", "");
                    }
                    var regexpContenType = /^contenttype\-/;
                    if (regexpContenType.test(value)) {
                        aContentTypeClass = value.replace("contenttype-", "");
                    }
                });
         
                // Check the right CSS class to be given to the main element
                if (main_elem.children(":last").is($.collective_navigationtoggle.listType)) {
                    main_elem.addClass('cnavOpen');
                } else {
                    main_elem.addClass('cnavClosed');
                } 
                control.bind("click",
                        {main_elem: main_elem, ul_model: ul_model, li_model: li_model,
                         control: control, wrapDiv: wrapDiv,
                         reviewStateClass: aReviewStateClass,  contentTypeClass: aContentTypeClass
                        },
                        checkClick);
    
            });
        };
    
    
        /**
         * Simple function that perform the whole check for every possible navigation elements
         * needing toggle feature.
         */
        var checkDOM = function () {
            $.each($.collective_navigationtoggle.toggle_elements, makeDynamicElements);
        };
    
    
        /**
         * Click handler to be used on expasible/collapsible navigation elements.
         * @param {Event} event jQuery event object, containing the data structure
         */
        var checkClick = function (event) {
            var new_ul;
            // Load data from event
            var main_elem = event.data.main_elem;
            var ul_model = event.data.ul_model;
            var li_model = event.data.li_model;
            var wrapDiv = event.data.wrapDiv;
            var control = event.data.control;
            // optional element classes
            var reviewStateClass = event.data.reviewStateClass;
            var contentTypeClass = event.data.contentTypeClass;
            // cache?
            var cache = $.collective_navigationtoggle.cache;
            
            if (main_elem.hasClass('cnavClosed')) {
                // check if the subtree is in the cache
                if (cache && main_elem.data('cnavCache')) {
                    new_ul = main_elem.data('cnavCache');
                    main_elem.append(new_ul);
                    checkDOM();
                    if ($.collective_navigationtoggle.slide_animation > 0) {
                        new_ul.slideDown($.collective_navigationtoggle.slide_animation, function () {
                            main_elem.removeClass('cnavClosed').addClass('cnavOpen');
                        });                    
                    } else {
                        main_elem.removeClass('cnavClosed').addClass('cnavOpen');
                        new_ul.show();
                    }
                } else {
                    main_elem.removeClass('cnavClosed').addClass('cnavOpen');
                    new_ul = ul_model.clone().hide();
                    main_elem.append(new_ul);
                    riseNavigatorClass(new_ul);
                    // Fill the new element
                    $.getJSON(call_context + '@@query-subelements',
                            {'path': $('a', main_elem).attr('href'),
                             'foo': loading_time}, // for cache prevention management
                            function (data) {
                                $.each(data, function (index, value) {
                                    new_ul.append(makeSubelement(value, li_model.clone(), value.icon,
                                                                 wrapDiv, reviewStateClass, contentTypeClass));
                                });
                                // If no element returned from the subtree, perform normal browser action
                                if ($($.collective_navigationtoggle.listItem, new_ul).length === 0) {
									var static_href = null;
									if (control.is('a')) {
										static_href = control.attr('href');
									} else {
										// in the control is not on a link, let's twy to see in is contained in a link
										static_href = control.closest('a').attr('href');
									}
                                    if (static_href) {
										$.collective_navigationtoggle._goTo(static_href);
									}
                                    return;
                                }
                                // I call this to obtain something like jQuery.live() feature...
                                checkDOM();
                                // Caching for later clicks
                                if (cache) {
                                    // As far as I'm not sure to rely on jQuery 1.4, I can't use the clone() withDataAndEvents
                                    main_elem.data('cnavCache', new_ul.clone(false));                                
                                }
                                // effects?
                                if ($.collective_navigationtoggle.slide_animation > 0) {
                                    new_ul.slideDown($.collective_navigationtoggle.slide_animation);
                                } else {
                                    new_ul.show();
                                }
                            }
                    );
                }
                event.preventDefault();                
            }
            else if (main_elem.hasClass('cnavOpen')) {
                new_ul = main_elem.children(":last");
                if ($.collective_navigationtoggle.slide_animation > 0) {
                    new_ul.slideUp($.collective_navigationtoggle.slide_animation, function () {
                        new_ul.remove();
                        main_elem.removeClass('cnavOpen').addClass('cnavClosed');
                    });                
                } else {
                    new_ul.remove();
                    main_elem.removeClass('cnavOpen').addClass('cnavClosed');
                }
                
                event.preventDefault();
            }
        };

        checkDOM();
		
		$(document).bind('checkDOM.navigationtoggle', function() {
			checkDOM();
		})

    });
})(jQuery);
